### PR TITLE
staging bucket and content manager refactor - part 1

### PR DESCRIPTION
**Pull request by en0ma**
This is part 1/3 of the staging bucket and content manager refactor. The idea is to make the staging bucket and content manager implementation more intuitive with deterministic behavior. 

It will also add more operator control by moving the staging bucket and content manager parameter constants to the config and making them configurable via CLI flags.

While allowing operators to decide their defaults. Part 2/3 will validate that operator values like deal duration, deal sizes, etc, are with the Filecoin network params expectations.

closes https://github.com/application-research/estuary/issues/367

### CODE REVIEW

File: handlers.go [hunk 1]
```diff
@@ -781,7 +781,7 @@ func (s *Server) handleAddCar(c echo.Context, u *User) error {
 
 	go func() {
 		// TODO: we should probably have a queue to throw these in instead of putting them out in goroutines...
-		s.CM.ToCheck <- cont.ID
+		s.CM.toCheck(cont.ID)
```
Comment (gmelodie) [inline handlers.go:784]: I think we should name this `checkContent` or maybe `check` if this is going to be a function

File: handlers.go [hunk 2]
```diff
@@ -943,7 +943,7 @@ func (s *Server) handleAdd(c echo.Context, u *User) error {
 	}
 
 	go func() {
-		s.CM.ToCheck <- content.ID
+		s.CM.toCheck(content.ID)
```
Comment (gmelodie) [inline handlers.go:946]: Same here, adjust naming

File: handlers.go [hunk 3]
```diff
@@ -1179,7 +1179,7 @@ func (s *Server) handleEnsureReplication(c echo.Context) error {
 
 	fmt.Println("Content: ", content.Cid.CID, data)
 
-	s.CM.ToCheck <- content.ID
+	s.CM.toCheck(content.ID)
```
Comment (gmelodie) [inline handlers.go:1182]: Adjust function name

File: pinning.go [hunk 4]
```diff
@@ -142,7 +142,7 @@ func (s *Server) doPinning(ctx context.Context, op *pinner.PinningOperation, cb
 	}
 
 	if op.MakeDeal {
-		s.CM.ToCheck <- op.ContId
+		s.CM.toCheck(op.ContId)
```
Comment (gmelodie) [inline pinning.go:145]: Adjust function name

File: config/deal.go [hunk 5]
```diff
@@ -1,7 +1,10 @@
...
 
+import "github.com/filecoin-project/go-state-types/abi"
+
 type Deal struct {
-	FailOnTransferFailure bool `json:"fail_on_transfer_failure"`
-	Disable               bool `json:"disable"`
-	Verified              bool `json:"verified"`
+	FailOnTransferFailure bool           `json:"fail_on_transfer_failure"`
+	Disabled              bool           `json:"disabled"`
+	Verified              bool           `json:"verified"`
```
Comment (gmelodie) [inline config/deal.go:8]: Nit: I like `IsVerified` and `IsDisabled` better

File: pinning.go [hunk 6]
```diff
@@ -955,7 +955,6 @@ func (cm *ContentManager) handlePinningComplete(ctx context.Context, handle stri
 		return xerrors.Errorf("failed to add objects to database: %w", err)
 	}
 
-	cm.ToCheck <- cont.ID
-
+	cm.toCheck(cont.ID)
```
Comment (gmelodie) [inline pinning.go:958]: Adjust function name

File: replication.go [hunk 7]
```diff
@@ -3331,10 +3301,10 @@ func (cm *ContentManager) splitContentLocal(ctx context.Context, cont util.Conte
 			return err
 		}
 
-		//queue splited contents
+		// queue splited contents
 		go func() {
-			log.Infow("queuing splited content child", "parent_contID", cont.ID, "child_contID", content.ID)
-			cm.ToCheck <- content.ID
+			log.Debugw("queuing splited content child", "parent_contID", cont.ID, "child_contID", content.ID)
+			cm.toCheck(content.ID)
```
Comment (gmelodie) [inline replication.go:3307]: Adjust function name

File: replication.go [hunk 8]
```diff
@@ -322,78 +302,101 @@ func NewContentManager(db *gorm.DB, api api.Gateway, fc *filclient.FilClient, tb
...
+	if !cm.localContentAddingDisabled {
+		go func() {
+			if err := cm.refreshPinQueue(ctx, util.ContentLocationLocal); err != nil {
+				log.Errorf("failed to refresh pin queue: %s", err)
+			}
+		}()
 	}
 
-	timer := time.NewTimer(time.Minute * 5)
+	// if staging buckets is enabled, rebuild the buckets, and run the bucket aggregate worker
```
Comment (gmelodie) [inline replication.go:338]:
```suggestion
	// if staging buckets are enabled, rebuild the buckets, and run the bucket aggregate worker
```

File: replication.go [hunk 9]
```diff
@@ -322,78 +302,101 @@ func NewContentManager(db *gorm.DB, api api.Gateway, fc *filclient.FilClient, tb
...
+	if cm.cfg.StagingBucket.Enabled {
+		// rebuild the staging buckets
```
Comment (gmelodie) [inline replication.go:340]: :scissors:

File: replication.go [hunk 10]
```diff
@@ -322,78 +302,101 @@ func NewContentManager(db *gorm.DB, api api.Gateway, fc *filclient.FilClient, tb
...
+		if err := cm.reBuildStagingBuckets(); err != nil {
```
Comment (gmelodie) [inline replication.go:341]:
```suggestion
		if err := cm.rebuildStagingBuckets(); err != nil {
```

File: replication.go [hunk 11]
```diff
@@ -322,78 +302,101 @@ func NewContentManager(db *gorm.DB, api api.Gateway, fc *filclient.FilClient, tb
...
+					for _, b := range buckets {
+						if err := cm.aggregateContent(context.TODO(), b); err != nil {
+							log.Errorf("content aggregation failed (bucket %d): %s", b.ContID, err)
+							continue
+						}
+					}
+					timer.Reset(time.Minute * 5)
+				}
 			}
+		}()
```
Comment (gmelodie) [inline replication.go:363]: Put this into a named function

File: replication.go [hunk 12]
```diff
@@ -322,78 +302,101 @@ func NewContentManager(db *gorm.DB, api api.Gateway, fc *filclient.FilClient, tb
...
-	for {
-		select {
-		case c := <-cm.ToCheck:
-			var content util.Content
-			if err := cm.DB.First(&content, "id = ?", c).Error; err != nil {
-				log.Errorf("finding content %d in database: %s", c, err)
-				continue
+		// run the bucket aggregator worker
+		go func() {
+			timer := time.NewTimer(time.Minute * 5)
```
Comment (gmelodie) [inline replication.go:347]: This should be a `const` var

File: replication.go [hunk 13]
```diff
@@ -322,78 +302,101 @@ func NewContentManager(db *gorm.DB, api api.Gateway, fc *filclient.FilClient, tb
...
+			for {
+				select {
+				case <-timer.C:
+					log.Debugw("content check queue", "length", len(cm.queueMgr.queue.elems), "nextEvent", cm.queueMgr.nextEvent)
+
+					buckets := cm.popReadyStagingZone()
+					for _, b := range buckets {
+						if err := cm.aggregateContent(context.TODO(), b); err != nil {
```
Comment (gmelodie) [inline replication.go:355]: We use `context.Background` more often on the code and [`context.Background()` and `context.TODO()` return the same context object](https://www.mongodb.com/community/forums/t/go-why-context-todo-and-not-context-background/5009/2) so let's just stick with Background
```suggestion
						if err := cm.aggregateContent(context.Background(), b); err != nil {
```

File: replication.go (hunk 11 above)
Comment (gmelodie) [inline replication.go:363]: It seems that we want to execute this every 5 minutes. In this case we should be using a [`Ticker`](https://gobyexample.com/tickers) and not a [`Timer`](https://gobyexample.com/timers)

File: replication.go [hunk 14]
```diff
@@ -655,7 +656,7 @@ func (cm *ContentManager) aggregateContent(ctx context.Context, b *contentStagin
 		}
 
 		go func() {
-			cm.ToCheck <- b.ContID
+			cm.toCheck(b.ContID)
```
Comment (gmelodie) [inline replication.go:659]: adjust naming

File: replication.go [hunk 15]
```diff
@@ -684,16 +686,11 @@ func (cm *ContentManager) createAggregate(ctx context.Context, conts []util.Cont
...
 	return dir, nil
 }
 
-func (cm *ContentManager) startup() error {
-	return cm.queueAllContent()
-}
-
-func (cm *ContentManager) reBuildStagingZones() error {
-	log.Info("rebuilding staging zones.......")
+func (cm *ContentManager) reBuildStagingBuckets() error {
```
Comment (gmelodie) [inline replication.go:692]:
```suggestion
func (cm *ContentManager) rebuildStagingBuckets() error {
```

File: replication.go [hunk 16]
```diff
@@ -1205,96 +1194,59 @@ func (cm *ContentManager) popReadyStagingZone() []*contentStagingZone {
...
 	}
 
+	// If this content is already scheduled to be aggregated and is waiting in a bucket
 	if cm.contentInStagingZone(ctx, content) {
-		// This content is already scheduled to be aggregated and is waiting in a bucket
 		return nil
 	}
 
-	// its too big, need to split it up into chunks
+	// its too big, need to split it up into chunks, no need to requeue dagsplit root content
```
Comment (gmelodie) [inline replication.go:1228]:
```suggestion
	// it's too big, need to split it up into chunks, no need to requeue dagsplit root content
```

File: replication.go [hunk 17]
```diff
@@ -1205,96 +1194,59 @@ func (cm *ContentManager) popReadyStagingZone() []*contentStagingZone {
...
-
-	minersAlready := make(map[address.Address]bool)
-	for _, d := range deals {
-		if d.Failed {
-			// TODO: this is an interesting choice, because it gives miners more chances to try again if they fail.
-			// I think that as we get a more diverse set of stable miners, we can *not* do this.
-			continue
+	// if staging bucket is enabled, try to bucket the content
+	if cm.canStageContent(content) {
+		// Only contents that has no deals and that are not themselves buckets(aggregates) that can be placed in a bucket
```
Comment (gmelodie) [inline replication.go:1246]:
```suggestion
		// Only contents that have no deals and that are not themselves buckets(aggregates) can be placed in a bucket
```

File: replication.go [hunk 18]
```diff
@@ -1353,7 +1306,14 @@ func (cm *ContentManager) ensureStorage(ctx context.Context, content util.Conten
...
 
 	goodDeals := numSealed + numPublished + numProgress
-	if goodDeals < replicationFactor {
+
+	replicationFactor := cm.Replication
+	if content.Replication > 0 {
+		replicationFactor = content.Replication
+	}
+
+	dealsTobeMade := replicationFactor - goodDeals
```
Comment (gmelodie) [inline replication.go:1315]:
```suggestion
	dealsToBeMade := replicationFactor - goodDeals
```

File: replication.go [hunk 19]
```diff
@@ -1353,7 +1306,14 @@ func (cm *ContentManager) ensureStorage(ctx context.Context, content util.Conten
...
+	if dealsTobeMade > 0 {
```
Comment (gmelodie) [inline replication.go:1316]:
```suggestion
	if dealsToBeMade > 0 {
```

File: replication.go [hunk 20]
```diff
@@ -1405,10 +1362,25 @@ func (cm *ContentManager) ensureStorage(ctx context.Context, content util.Conten
...
+				return err
+			}
+			excludedMiners[maddr] = true
+		}
+
 		go func() {
 			// make some more deals!
-			log.Infow("making more deals for content", "content", content.ID, "curDealCount", len(deals), "newDeals", replicationFactor-len(deals))
-			if err := cm.makeDealsForContent(ctx, content, replicationFactor-len(deals), minersAlready, verified); err != nil {
+			log.Infow("making more deals for content", "content", content.ID, "curDealCount", len(deals), "newDeals", dealsTobeMade)
```
Comment (gmelodie) [inline replication.go:1382]:
```suggestion
			log.Infow("making more deals for content", "content", content.ID, "curDealCount", len(deals), "newDeals", dealsToBeMade)
```

File: replication.go [hunk 21]
```diff
@@ -1405,10 +1362,25 @@ func (cm *ContentManager) ensureStorage(ctx context.Context, content util.Conten
...
+			if err := cm.makeDealsForContent(ctx, content, dealsTobeMade, excludedMiners); err != nil {
```
Comment (gmelodie) [inline replication.go:1383]:
```suggestion
			if err := cm.makeDealsForContent(ctx, content, dealsToBeMade, excludedMiners); err != nil {
```

File: replication.go (hunk 19 above)
Comment (gmelodie) [inline replication.go:1316]: Let's do the following instead:
```go
if dealsToBeMade <= 0 {
    done(time.Minute * 10)
    return nil
}

Comment (gmelodie) [reply replication.go:1316]: Also add `time.Minute * 10` as a const var

File: replication.go [hunk 22]
```diff
@@ -1370,7 +1330,6 @@ func (cm *ContentManager) ensureStorage(ctx context.Context, content util.Conten
 					done(time.Second * 10)
```
Comment (gmelodie) [inline replication.go:1330]: Make this a const var

File: replication.go [hunk 23]
```diff
@@ -1379,10 +1338,8 @@ func (cm *ContentManager) ensureStorage(ctx context.Context, content util.Conten
 				if err := cm.RefreshContent(context.Background(), content.ID); err != nil {
 					log.Errorf("failed to retrieve content in need of repair %d: %s", content.ID, err)
 				}
-
 				done(time.Second * 30)
```
Comment (gmelodie) [inline replication.go:1341]: Make this a const var

File: replication.go [hunk 24]
```diff
@@ -1405,10 +1362,25 @@ func (cm *ContentManager) ensureStorage(ctx context.Context, content util.Conten
...
+
 		go func() {
 			// make some more deals!
-			log.Infow("making more deals for content", "content", content.ID, "curDealCount", len(deals), "newDeals", replicationFactor-len(deals))
-			if err := cm.makeDealsForContent(ctx, content, replicationFactor-len(deals), minersAlready, verified); err != nil {
+			log.Infow("making more deals for content", "content", content.ID, "curDealCount", len(deals), "newDeals", dealsTobeMade)
+			if err := cm.makeDealsForContent(ctx, content, dealsTobeMade, excludedMiners); err != nil {
 				log.Errorf("failed to make more deals: %s", err)
 			}
 			done(time.Minute * 10)
```
Comment (gmelodie) [inline replication.go:1386]: Make this a const var

**Review by gmelodie — CHANGES_REQUESTED** [review]
Good refactor!

File: config/estuary.go [hunk 25]
```diff
@@ -68,16 +71,29 @@ func NewEstuary(appVersion string) *Estuary {
 		EnableAutoRetrieve:     false,
 
 		Deal: Deal{
-			Disable:               false,
+			Disabled:              false,
 			FailOnTransferFailure: false,
 			Verified:              true,
+			Duration:              abi.ChainEpoch(1555200 - (2880 * 21)), // Making default deal duration be three weeks less than the maximum to ensure miners who start their deals early dont run into issues
```
[the remark this answers is not in the document]
Comment (en0ma) [reply config/estuary.go:77]: these will come in part 2

File: config/estuary.go [hunk 26]
```diff
@@ -68,16 +71,29 @@ func NewEstuary(appVersion string) *Estuary {
...
+			Enabled:                 true,
+			MaxLifeTime:             time.Hour * 8,
+			MaxContentAge:           time.Hour * 24 * 7,
+			MaxItems:                10000,
+			MaxSize:                 int64((abi.PaddedPieceSize(16<<30).Unpadded() * 9) / 10),                // 14.29 Gib
+			MinSize:                 int64(int64((abi.PaddedPieceSize(16<<30).Unpadded()*9)/10) - (1 << 30)), // 13.29 GiB
+			KeepAlive:               time.Minute * 40,
+			MinDealSize:             256 << 20,                                               //0.25 Gib
+			IndividualDealThreshold: int64((abi.PaddedPieceSize(4<<30).Unpadded() * 9) / 10), // 90% of the unpadded data size for a 4GB piece, the 10% gap is to accommodate car file packing overhead, can probably do this better
+		},
```
[the remark this answers is not in the document]
Comment (en0ma) [reply config/estuary.go:95]: these will come in part 2

File: handlers.go (hunk 1 above)
Comment (en0ma) [reply handlers.go:784]: these will come in part 2

File: handlers.go (hunk 2 above)
Comment (en0ma) [reply handlers.go:946]: these will come in part 2

File: replication.go (hunk 23 above)
Comment (en0ma) [reply replication.go:1341]: these will come in part 2

File: replication.go (hunk 24 above)
Comment (en0ma) [reply replication.go:1386]: these will come in part 2

File: pinning.go (hunk 4 above)
Comment (en0ma) [reply pinning.go:145]: will address in part 2

File: handlers.go (hunk 3 above)
Comment (en0ma) [reply handlers.go:1182]: will address in part 2

File: pinning.go (hunk 6 above)
Comment (en0ma) [reply pinning.go:958]: will address in part 2

File: replication.go (hunk 7 above)
Comment (en0ma) [reply replication.go:3307]: will address in part 2

File: replication.go (hunk 9 above)
Comment (en0ma) [reply replication.go:340]: @gmelodie what do you mean?

**Review by gmelodie — DISMISSED** [review]

**Review by gmelodie — DISMISSED** [review]
Looking good! @en0ma just fix the conflicts and I think we can merge this on `dev`

**Review by gmelodie — APPROVED** [review]

**Pull request merged**